Fixsubstitutions

adinn opened the pull request.

adinn commented:
This fix corrects several errors that combine to select the wrong source file for a substituted method or field. It modifies DebugTypeInfo and DebugCodeInfo to ensure that the annotated class is always used when trying to find the file for a substituted method or field (in one of the two cases where this went wrong the breakage was caused by a change to SubstitutionMethod/Field).  It also ensures that the line info generator uses starts off using the file associated with a primary address range/method rather than defaulting to the original class. In the case of a method/field substitution this ensures the file of the annotated class is used.

The patch also includes a commit to the debuginfotest gate check (provided by @zakkak) that ensures gdb locates a substituted method in the correct file when it is broken. This change tests the current fix and should also help to catch any other changes that might break lookup of the substituted method.

adinn commented:
@olpaw This fix is ready for review if you would be so kind as to look at it.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLineSectionImpl.java:
```diff
@@ -425,24 +425,30 @@ private int writeFileTable(DebugContext context, ClassEntry classEntry, byte[] b
         int pos = p;
-        FileEntry fileEntry = classEntry.getFileEntry();
-        if (fileEntry == null) {
+        // if the class has not associated file then don't generate line onfo
```
s/onfo/info

adinn replied:
fixed

olpaw commented on substratevm/src/com.oracle.svm.test/src/hello/Target_hello_Hello_DefaultGreeter.java:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
```
2021, 2021

adinn replied:
fixed

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -226,14 +227,17 @@ protected static ResolvedJavaType getJavaType(HostedField hostedField, boolean w
+                return substitutionField.getAnnotated().getDeclaringClass();
+            }
+        }
         return javaField.getDeclaringClass();
```
I would rename those `getJavaType` methods to `getDeclaringClass`. IMO `getJavaType` is too unspecific (doesn't tell us that we want the type where the method/field is declared in).

adinn replied:
Done

olpaw reviewed (changes requested):
Please rename `com.oracle.svm.hosted.image.NativeImageDebugInfoProvider#getJavaType` as explained. Other than that perfectly fine.

zakkak commented:
Hi @adinn, thanks for the fix. 

Please note that this PR is not based on the latest master and is thus not applied on top of #3304. The PR can be cleanly rebased but what concerns me is the following. In #3304 we associate each range with a `MethodEntry` and each `MethodEntry` is associated with a `FileEntry`, all the `MethodEntry`s corresponding to a Class' methods are kept in the `methods` field of `ClassEntry`. The creation of (most of) the `MethodEntry`s is done by visiting all the methods returned by `getAllDeclaredMethods()` of the corresponding type/Class. @olpaw is there any chance for `getAllDeclaredMethods()` to return a `HostedMethod` of the original method instead of a `HostedMethod` wrapping a `SubstitutionMethod`?

Furthermore just to make sure that I have a clear picture.

Leaving https://github.com/adinn/graal/blob/316ff62d918afd9850d40d6e753f404365eadd6e/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java#L685 as is is intentional, right?

This way you ensure that `info functions hello.Hello$DefaultGreeter::greet` shows greet under `hello/Hello.java` while `breakpoint`, `list` and `backtrace` show the right source location (under `hello/Target_hello_Hello_DefaultGreeter.java`).

zakkak commented:
This PR fixes #3420

adinn commented:
> Please note that this PR is not based on the latest master and is thus not applied on top of #3304. The PR can be cleanly rebased but what concerns me is the following.

Yes, apologies for that -- I thought I had pulled in all upstream changes. I have rebased and corrected the branch.

> In #3304 we associate each range with a MethodEntry and each MethodEntry is associated with a FileEntry, all the MethodEntrys corresponding to a Class' methods are kept in the methods field of ClassEntry. The creation of (most of) the MethodEntrys is done by visiting all the methods returned by getAllDeclaredMethods() of the corresponding type/Class. @olpaw is there any chance for getAllDeclaredMethods() to return a HostedMethod of the original method instead of a HostedMethod wrapping a SubstitutionMethod?

There is no need to do that. Indeed it would be wrong to do so. The MethodEntry determines its file using logic provided in the NativeImageDebugInfo class hierarchy. This logic was broken but I have just corrected it. When a MethodEntry is derived from a SubstitutionMethod then NativeImageDebugInfo will derive its file name using the annotated class - -either using the filename stored in the bytecode or using the class name. It will derive the method owner name using the original class. When it is derived from a non-substitution Method both filename and class name are derived from the original class.

> Furthermore just to make sure that I have a clear picture.

> Leaving https://github.com/adinn/graal/blob/316ff62d918afd9850d40d6e753f404365eadd6e/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java#L685 as is is intentional, right?

> This way you ensure that info functions hello.Hello$DefaultGreeter::greet shows greet under hello/Hello.java while breakpoint, list and backtrace show the right source location (under hello/Target_hello_Hello_DefaultGreeter.java).

No, that is not intentional. It is a mistake. When a function is listed it should be listed via the file it's source code belongs to not the file that it's class belongs to. I'll post a fix for this in this PR.

olpaw commented:
@adinn please rebase once again so that we get rid of the merge conflict.

adinn commented:
@olpaw I have rebased over master.

I also fixed the file problem @zakkak noted with info listings and I updated the debuginfotest to check that the substituted method is indeed listed as belonging to the target class file.

So, this ought to be complete and ready to integrate.

adinn commented:
@olpaw How can I disable this style check? 

The check is clearly inappropriate here. The substituted method cannot actaully be converted to a static as it is substituting an instance method. Do I really have to make the code use `this` to shut the checker up or is there a way to disable the check? I tried moving the checkstyls stop and start comments around the method as a whole but the check is still being enforced.

olpaw commented:
> @olpaw How can I disable this style check?
> 
> The check is clearly inappropriate here. The substituted method cannot actaully be converted to a static as it is substituting an instance method. Do I really have to make the code use `this` to shut the checker up or is there a way to disable the check? I tried moving the checkstyls stop and start comments around the method as a whole but the check is still being enforced.

Please try if `@SuppressWarnings("static-method")` does the trick.

adinn commented:
@olpaw Thanks!

olpaw commented:
Running in our CI now.

olpaw commented:
All gates **but one** passed our CI. It's a Windows gate that builds an image with debuginfo. There we see the following
```
Caused by: java.lang.NullPointerException
	at com.oracle.objectfile.debugentry.DebugInfoBase.lambda$installDebugInfo$5(DebugInfoBase.java:249)
	at com.oracle.svm.hosted.image.NativeImageDebugInfoProvider$NativeImageDebugCodeInfo.debugContext(NativeImageDebugInfoProvider.java:883)
	at com.oracle.objectfile.debugentry.DebugInfoBase.lambda$installDebugInfo$6(DebugInfoBase.java:234)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.TreeMap$EntrySpliterator.forEachRemaining(TreeMap.java:2962)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at com.oracle.objectfile.debugentry.DebugInfoBase.installDebugInfo(DebugInfoBase.java:234)
	at com.oracle.objectfile.pecoff.PECoffObjectFile.installDebugInfo(PECoffObjectFile.java:726)
	at com.oracle.svm.hosted.image.NativeImage.build(NativeImage.java:456)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:682)
	... 7 more
```
It would be interesting if this is also reproduceable on your side. Maybe @stooke can take a look if he can reproduce it on his Windows machine. I'm also investigating further ... once my Windows VM has completed its updating galore :roll_eyes:

stooke commented:
I can try it - is there a pointer to the file being compiled?

zakkak commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java:
```diff
@@ -682,7 +682,8 @@ private int writeMethodDeclaration(DebugContext context, ClassEntry classEntry,
-        int fileIdx = classEntry.localFilesIdx();
+        FileEntry fileEntry = range.getFileEntry();
+        int fileIdx = (fileEntry != null ? classEntry.localFilesIdx(fileEntry) : classEntry.localFilesIdx());
         log(context, "  [0x%08x]     file 0x%x (%s)", pos, fileIdx, range.getFileEntry().getFullName());
```
`range.getFileEntry()` may be null here.

adinn replied:
Hi Foivos.
This is currently going through the gate so it would be better to fix this in a follow-up patch.
Perhaps you can fold this into your patch for #3419

zakkak replied:
ACK, it's already fixed there (that's how I noticed it in the first place)!

olpaw commented:
This PR is merged but mergebot was not able to detect this due to added commits after approval.
See https://github.com/oracle/graal/commit/0242b5ea1f32b15ade620408dfdee4b9b0b66746
